### PR TITLE
Master mdl 71696 versioning mod quiz 193 bhtfix 268

### PR DESCRIPTION
**Pull request by marcghaly**
Behat `core_question`, `qbank` tags are passing properly.
`mod_quiz` has 2 failing test scenarios which are the following: 
- `001 Scenario: Restore a Moodle 2.8 quiz backup`
- `002 Scenario: Dry-run a full regrade, then regrade the attempts that will need it.`

First scenario belongs to another branch and second needs discussion.

ps: Some changes made may belong to other branches.

### CODE REVIEW

File: lib/tests/behat/behat_general.php [hunk 1]
```diff
@@ -634,6 +634,57 @@ function($context, $args) {
...
 
+    /**
+     * Checks, that page contains specified text or another. It also checks if the text is visible when running Javascript tests.
+     *
+     * @Then /^I should see "(?P<text_string>(?:[^"]|\\")*)" or "(?P<text_string2>(?:[^"]|\\")*)"$/
+     * @throws ExpectationException
+     * @param string $text
+     * @param string $secondtext
+     */
+    public function i_should_see_or($text, $secondtext) {
```
Comment (guillogo) [inline lib/tests/behat/behat_general.php:645]: @marcghaly I'm not sure about touching this ... as `lib/tests/behat/behat_general.php ` core behat library. Adding code here seems for me not right.

Comment (safatshahin) [reply lib/tests/behat/behat_general.php:645]: please revert any changes here, this is core behat library

File: mod/quiz/locallib.php [hunk 2]
```diff
@@ -2174,11 +2174,17 @@ function quiz_require_question_use($questionid) {
...
     global $DB;
-    $question = $DB->get_record_sql("
-            SELECT q.*
+
+    $sql = 'SELECT q.*
               FROM {quiz_slots} slot
-              JOIN {question} q ON q.id = slot.questionid
-             WHERE slot.quizid = ? AND slot.slot = ?", array($quiz->id, $slot));
+              JOIN {question_references} qre ON slot.id = qre.itemid
+              JOIN {question_bank_entry} qbe ON qre.questionbankentryid = qbe.id
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply mod/quiz/locallib.php:2181]: done

File: mod/quiz/tests/behat/behat_mod_quiz.php [hunk 3]
```diff
@@ -247,7 +252,7 @@ public function quiz_contains_the_following_questions($quizname, TableNode $data
                 } else {
                     $includingsubcategories = clean_param($questiondata['includingsubcategories'], PARAM_BOOL);
                 }
-                quiz_add_random_questions($quiz, $page, $question->category, 1, $includingsubcategories);
+                quiz_add_random_questions($quiz, $page, $question->questioncategoryid, 1, $includingsubcategories);
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply mod/quiz/tests/behat/behat_mod_quiz.php:255]: done

File: lib/tests/behat/behat_general.php (hunk 1 above)
Comment (marcghaly) [reply lib/tests/behat/behat_general.php:645]: moved to `mod/quiz/tests/behat/behat_mod_quiz.php`

File: mod/quiz/classes/question/bank/qbank_helper.php [hunk 4]
```diff
@@ -276,7 +276,7 @@ public static function get_question_report_structure_data($quizid, $questionids
             $condition = 'AND q.id ' . $condition;
             $params = array_merge($params, $param);
         }
-        $selectstart = 'slot.slot, q.id,';
+        $selectstart = 'q.id, slot.slot,';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply mod/quiz/classes/question/bank/qbank_helper.php:279]: That change has been reverted as for point 2, not every grade were displayed. Problem is, slot is not unique and is triggering a debugging message in behat.

**Comment by safatshahin** [comment]
Hi @marcghaly 
Thanks for working in this, just some points:
1. We need some behat coverage for versioning.
2. Please remove any leftover code.
3. Cheers

**Pull request closed without merging**